src/io_threads.c

```diff
@@ -353,17 +936,121 @@ void initIOThreads(void) {
     serverAssert(server.io_threads_num <= IO_THREADS_MAX_NUM);
 
     prefetchCommandsBatchInit();
+    size_t io_to_mt_queue_size = (server.io_threads_num - 1) * DEFAULT_MPSC_QUEUE_SIZE_PER_THREAD;
+    io_to_mt_queue = IoToMTQueueCreate(io_to_mt_queue_size);
+    deferredCmdExclusive.pending_clients = listCreate();
+    deferredCmdExclusive.deferred_jobs = listCreate();
 
     /* Spawn and initialize the I/O threads. */
     for (int i = 1; i < server.io_threads_num; i++) {
         createIOThread(i);
     }
 }
 
+/*
+ * This function is called when a client is closed but still has pending IO jobs.
+ * It tracks the client in a dictionary in order to know to ignore its pending jobs.
+ */
+void ioThreadsOnUnlinkClient(client *c) {
+    if (c->bstate && c->bstate->slot_pending_list) {
+        dqRemoveClient(c->bstate->slot_pending_list, c);
+        c->bstate->slot_pending_list = NULL;
+    }
+}
+
+/*  returns C_OK if the command is postpone due to busy slot */
+static int isCommandPostpone(client *c) {
+    deferredQueue *q = getDeferredQueue(c->slot);
+    if (!dqAvailable(q)) {
+        dqAddPendingClient(q, c);
+        return C_OK; /* Postpone the command execution */
+    }
+    return C_ERR;
+}
+
+/* Check if a command can be offloaded to IO threads.
+ * Returns 1 if the command can be offloaded, 0 otherwise. */
+int canCommandBeOffloaded(int slot, struct serverCommand *cmd) {
+    if (slot == -1) return 0;
+
+    if (!server.cluster_enabled) {
+        return 0; /* Avoid offloading commands in non cluster mode. */
+    }
+
+    if (server.active_io_threads_num == 1) {
+        return 0; /* No IO threads to offload to. */
+    }
+
+    if (!server.io_threads_do_commands_offloading) {
+        return 0; /* Command offloading is disabled. */
+    }
+
+    /* Check if modules are loaded and module offloading is disabled */
+    if (moduleCount() > 0 && !server.io_threads_do_commands_offloading_with_modules) {
```

Comment:
I'd imagine that a property access would be cheaper than a function call...
```suggestion
    if (!server.io_threads_do_commands_offloading_with_modules && moduleCount() > 0) {
```